msdsn/conn_str.go

```diff
@@ -74,21 +78,40 @@ type Config struct {
 	PacketSize  uint16
 }
 
-func SetupTLS(certificate string, insecureSkipVerify bool, hostInCertificate string) (*tls.Config, error) {
+func SetupTLS(certificate string, insecureSkipVerify bool, hostInCertificate string, allowNonstandardHostname bool) (*tls.Config, error) {
 	var config tls.Config
+	config.ServerName = hostInCertificate
 	if certificate != "" {
 		pem, err := ioutil.ReadFile(certificate)
 		if err != nil {
 			return nil, fmt.Errorf("cannot read certificate %q: %v", certificate, err)
 		}
-		certs := x509.NewCertPool()
-		certs.AppendCertsFromPEM(pem)
-		config.RootCAs = certs
-	}
-	if insecureSkipVerify {
-		config.InsecureSkipVerify = true
+		if !allowNonstandardHostname {
+			certs := x509.NewCertPool()
+			certs.AppendCertsFromPEM(pem)
+			config.RootCAs = certs
+			if insecureSkipVerify {
+				config.InsecureSkipVerify = true
+			}
+		} else {
+			// Set InsecureSkipVerify to skip the default hostname validation that this is
+			// replacing. This does not disable VerifyConnection.
+			config.InsecureSkipVerify = true
+			config.VerifyConnection = func(cs tls.ConnectionState) error {
+				commonName := cs.PeerCertificates[0].Subject.CommonName
```

Comment:
This verify Connection function doesn't open the "insecure Skip Verify"  bool. If insecure skip verify is true, then do an early return.